components/opponent/wikis/stormgate/opponent_custom.lua

```diff
@@ -0,0 +1,145 @@
+---
+-- @Liquipedia
+-- wiki=stormgate
+-- page=Module:Opponent/Custom
+--
+-- Please see https://github.com/Liquipedia/Lua-Modules to contribute
+--
+
+local Faction = require('Module:Faction')
+local Lua = require('Module:Lua')
+local String = require('Module:StringUtils')
+local Table = require('Module:Table')
+local TeamTemplate = require('Module:TeamTemplate')
+local TypeUtil = require('Module:TypeUtil')
+
+local Opponent = Lua.import('Module:Opponent', {requireDevIfEnabled = true})
+local PlayerExt = Lua.import('Module:Player/Ext/Custom', {requireDevIfEnabled = true})
+
+local CustomOpponent = Table.deepCopy(Opponent)
+
+CustomOpponent.types.Player = TypeUtil.extendStruct(Opponent.types.Player, {
+	faction = 'string?',
+})
+
+CustomOpponent.types.PartyOpponent = TypeUtil.struct{
+	players = TypeUtil.array(CustomOpponent.types.Player),
+	type = TypeUtil.literalUnion(unpack(Opponent.partyTypes)),
+}
+
+CustomOpponent.types.Opponent = TypeUtil.union(
+	Opponent.types.TeamOpponent,
+	CustomOpponent.types.PartyOpponent,
+	Opponent.types.LiteralOpponent
+)
+
+---@class StormgateStandardPlayer:standardPlayer
+---@field faction string?
+
+---@class StormgateStandardOpponent:standardOpponent
+---@field players StormgateStandardPlayer[]
+---@field isArchon boolean
+---@field isSpecialArchon boolean?
+---@field extradata table
+
+---@param args table
+---@return StormgateStandardOpponent?
+function CustomOpponent.readOpponentArgs(args)
+	local opponent = Opponent.readOpponentArgs(args) --[[@as StormgateStandardOpponent?]]
```

Comment:
```suggestion
	local opponent = Opponent.readOpponentArgs(args) ---@type StormgateStandardOpponent?
```

Reply:
Cannot assign `standardOpponent?` to `StormgateStandardOpponent?`.
- `standardOpponent` cannot match `StormgateStandardOpponent?`
- Type `standardOpponent` cannot match `nil`
- Type `standardOpponent` cannot match `StormgateStandardOpponent`